src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs

```diff
@@ -0,0 +1,226 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Threading;
+
+// <summary>
+// This class provides a way for browser threads to asynchronously wait for a semaphore
+// from JS, without using the threadpool.  It is used to implement threadpool workers.
+// </summary>
+internal sealed partial class LowLevelLifoAsyncWaitSemaphore : LowLevelLifoSemaphoreBase, IDisposable
+{
+    private IntPtr lifo_semaphore;
+
+    [MethodImplAttribute(MethodImplOptions.InternalCall)]
+    private static extern IntPtr InitInternal(int kind);
+
+    public LowLevelLifoAsyncWaitSemaphore(int initialSignalCount, int maximumSignalCount, int spinCount, Action onWait)
+        : base (initialSignalCount, maximumSignalCount, spinCount, onWait)
+    {
+        CreateAsyncWait(maximumSignalCount);
+    }
+
+#pragma warning disable IDE0060
+    private void CreateAsyncWait(int maximumSignalCount)
+#pragma warning restore IDE0060
+    {
+        lifo_semaphore = InitInternal(2);
+    }
+
+    [MethodImplAttribute(MethodImplOptions.InternalCall)]
+    private static extern void DeleteInternal(IntPtr semaphore);
+
+    public void Dispose()
+    {
+        DeleteInternal(lifo_semaphore);
+        lifo_semaphore = IntPtr.Zero;
+    }
+
+    [MethodImplAttribute(MethodImplOptions.InternalCall)]
+    private static extern void ReleaseInternal(IntPtr semaphore, int count);
+
+    protected override void ReleaseCore(int count)
+    {
+        ReleaseInternal(lifo_semaphore, count);
+    }
+
+    private sealed record WaitEntry (LowLevelLifoAsyncWaitSemaphore Semaphore, int TimeoutMs, Action<LowLevelLifoAsyncWaitSemaphore, object?> OnSuccess, Action<LowLevelLifoAsyncWaitSemaphore, object?> OnTimeout, object? State);
+
+    public void PrepareAsyncWait(int timeoutMs, Action<LowLevelLifoAsyncWaitSemaphore, object?> onSuccess, Action<LowLevelLifoAsyncWaitSemaphore, object?> onTimeout, object? state)
+    {
+        //FIXME(ak): the async wait never spins. Shoudl we spin a little?
+        Debug.Assert(timeoutMs >= -1);
```

Comment:
should we at least check whether the wait can complete instantly before entering the loop? or is that not a scenario that's possible for the lifo semaphore unlike other synchronization primitives